aip/0216.md

```diff
@@ -60,13 +60,67 @@ The field referencing the `State` enum in a resource **should** behave and be
 documented as "Output only", in accordance with [AIP-203](./0203.md).
 
 APIs **should not** allow a `State` enum to be directly updated through an
-"update" method, and **should** instead use a custom method.
+"update" method, and **should** instead use a custom state transition method.
 
 This is because update methods are generally not expected to have side effects,
 and also because updating state directly implies that it is possible to set the
 state to any available value, whereas states generally reflect a resource's
 progression through a lifecycle.
 
+### State transition methods
+
+State transition methods are a special type of custom method that are
+responsible for transitioning a state field from one enum value to another.
+As part of the transition, other fields may also change, e.g. an `update_time`
+field. The method definition should look like the following:
+
+```proto
+// Publishes a book.
+// The `state` of the book after publishing will be PUBLISHED.
+// `PublishBook` can be called on Books in the state DRAFT; Books in a
+// different state (including PUBLISHED) will return an error.
+rpc PublishBook(PublishBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:publish"
+    body: "*"
+  };
+}
+```
+
+- The name of the method **should** be a verb followed by the singular form of
+  the resource's message name.
+- The request message **must** match the RPC name, with a `-Request` suffix.
+- The response message **must** be the resource itself.
+  - If the create RPC is [long-running](#long-running-create), the response
+    message **must** be a `google.longrunning.Operation` which resolves to the
+    resource itself.
+- The HTTP verb **must** be `POST`.
+- The HTTP URI **must** use a `:` character followed by the custom verb
+  (`:publish` in the above example), and the verb in the URI **must** match the
+  verb in the name of the RPC.
+  - If word separation is required, `camelCase` **must** be used.
+- The `body` clause in the `google.api.http` annotation **must** be `"*"`.
+- The request message field receiving the resource name **should** map to the
+  URL path.
+  - This field **should** be called `name`.
+  - The `name` field **should** be the only variable in the URI path. All
+    remaining parameters **should** map to URI query parameters.
+- If the state transition is not allowed, the service **must** error with
+  `FAILED_PRECONDITION` (HTTP 400).
+
+The request message should look like this:
+
+```proto
+message PublishBookRequest {
+  // The name of the book to publish.
+  // Format: publishers/{publisher}/books/{book}
+  string name = 1;
+}
+```
+
+- A resource name field **must** be included. It **should** be called `name`.
+- The comment for the field **should** document the resource pattern.
```

Comment:
It is probably worth explicitly saying that other, arbitrary arguments are permitted.

Reply:
Added.